src/shared/transitions.js

```diff
@@ -4,37 +4,31 @@ export function linear(t) {
 	return t;
 }
 
-export function generateKeyframes(
+export function generateRule(
 	a,
 	b,
 	delta,
 	duration,
 	ease,
-	fn,
-	node,
-	style
+	fn
 ) {
-	var id = '__svelte' + ~~(Math.random() * 1e9); // TODO make this more robust
-	var keyframes = '@keyframes ' + id + '{\n';
+	var keyframes = '{\n';
 
 	for (var p = 0; p <= 1; p += 16.666 / duration) {
 		var t = a + delta * ease(p);
 		keyframes += p * 100 + '%{' + fn(t) + '}\n';
 	}
 
-	keyframes += '100% {' + fn(b) + '}\n}';
-	style.textContent += keyframes;
+	return keyframes + '100% {' + fn(b) + '}\n}';
+}
 
-	document.head.appendChild(style);
+// https://github.com/darkskyapp/string-hash/blob/master/index.js
+export function hash(str) {
+	var hash = 5381;
+	var i = str.length;
 
-	node.style.animation = (node.style.animation || '')
-		.split(',')
-		.filter(function(anim) {
-			// when introing, discard old animations if there are any
-			return anim && (delta < 0 || !/__svelte/.test(anim));
-		})
-		.concat(id + ' ' + duration + 'ms linear 1 forwards')
-		.join(', ');
+	while (i--) hash = (hash * 33) ^ str.charCodeAt(i);
```

Comment:
Well, whaddaya know — if you do the bitshifting yourself, you [get a significant speedup](https://esbench.com/bench/594eef9599634800a034870b)!

<img width="638" alt="screen shot 2017-06-24 at 7 34 39 pm" src="https://user-images.githubusercontent.com/1162160/27512605-336fc136-5914-11e7-8632-dcb6215cf4fc.png">

```js
function bitshift33 (str) {
  var hash = 5381;
  var i = str.length;

  while (i--) hash = ((hash << 5) + hash) ^ str.charCodeAt(i);
  return hash >>> 0;
}
```

I.e. it's a shift and addition rather than shift and subtraction.

According to [this](http://www.cse.yorku.ca/~oz/hash.html) (which https://github.com/darkskyapp/string-hash is based on) *33* does give better results than other numbers, including primes, and no-one is quite sure why. So I'll update this PR to use manual bitshifting, but I reckon we might be better off sticking with 33?